File: Headless/Headless/Ball.cs
using nkast.Aether.Physics2D.Common;
using System;
using System.Collections.Generic;
using System.Text;

namespace Headless
{
    internal class Ball
    {
        public Vector2 previousPosition;
        public Vector2 position;
        
        public readonly float radius;

        public Ball(Vector2 initPos, float radius)
        {
            this.position = initPos;
            this.radius = radius;
        }

        public Ball(float radius) 
        {
            this.position = Vector2.Zero;
            this.radius = radius;
        }

        public bool HasHitWall(Wall wall, bool actOnBothSides)
        {
            var toWallOriginPrevious = wall.origin - previousPosition;
            var distanceToWallPrevious = Helpers.Dot(toWallOriginPrevious, wall.orthogonalVector) - radius;
            var toWallOrigin = wall.origin - position;
            var distanceToWall = Helpers.Dot(toWallOrigin, wall.orthogonalVector) - radius;

            var inlineDistancePrevious = Helpers.Dot(toWallOriginPrevious, wall.inlineVector);
            var inlineDistance = Helpers.Dot(toWallOrigin, wall.inlineVector);

            //Console.WriteLine("To Wall Dist: " + distanceToWall);
            //Console.WriteLine("Inline Dist: " + inlineDistance);

            if (distanceToWall <= 0 && distanceToWallPrevious > 0 && Mathf.Abs(inlineDistance) < wall.length / 2)
            {
                return true;
            }
            if (actOnBothSides)
            {
                //var distanceToWallPreviousOtherSide = Helpers.Dot(toWallOriginPrevious, wall.reverseOrthogonalVector) - radius;
                //var distanceToWallOtherSide = Helpers.Dot(toWallOrigin, wall.reverseOrthogonalVector) - radius;
                //Console.WriteLine("To wall other side dist: " + distanceToWallOtherSide);
                if (distanceToWall >= 0 && distanceToWallPrevious < 0 && Mathf.Abs(inlineDistance) < wall.length / 2)
                {
                    return true;
                }
            }
            return false;
        }

        public bool HasHitWall(Wall wall)
        {
            return HasHitWall(wall, false);
        }

        public bool HasHitAnyWall(Wall[] walls, bool actOnBothSides)
        {
            foreach (var wall in walls)
            {
                if (HasHitWall(wall, actOnBothSides))
                {
                    return true;
                }
            }
            return false;
        }

        public bool HasHitAnyWall(Wall[] walls)
        {
            return HasHitAnyWall(walls, false);
        }
    }
}

File: Headless/Headless-rs/src/ball.rs
use crate::vec2::{self, Vec2};

pub struct Ball {
    pub previous_position: Vec2,
    pub position: Vec2,
    radius: f64
}

impl Ball {
    pub fn new(radius: f64) -> Self {
        Ball {
            previous_position: *vec2::ZERO,
            position: *vec2::ZERO,
            radius
        }
    }

    pub fn has_hit(&self, wall: Wall, act_on_both_sides: bool) -> bool {
        let to_wall_origin_previous = wall.origin - self.previous_position;
        let distance_to_wall_previous = to_wall_origin_previous.dot(wall.orthogonal) - self.radius;
        let to_wall_origin = wall.origin - self.position;
        let distance_to_wall = to_wall_origin.dot(wall.orthogonal) - self.radius;

        let inline_distance_previous = to_wall_origin_previous.dot(wall.inline);
        let inline_distance = to_wall_origin.dot(wall.inline);

        if distance_to_wall <= 0.0 && distance_to_wall_previous > 0.0 && inline_distance.abs() < wall.length / 2.0 {
            return true;
        }

        if act_on_both_sides {
            if distance_to_wall >= 0.0 && distance_to_wall_previous < 0.0 && inline_distance.abs() < wall.length / 2.0 {
                return true;
            }
        }

        return false;
    }

    pub fn has_hit_reverse(&self, wall: Wall) -> bool {
        let to_wall_origin_previous = wall.origin - self.previous_position;
        let distance_to_wall_previous = to_wall_origin_previous.dot(wall.orthogonal) - self.radius;
        let to_wall_origin = wall.origin - self.position;
        let distance_to_wall = to_wall_origin.dot(wall.orthogonal) - self.radius;

        let inline_distance_previous = to_wall_origin_previous.dot(wall.inline);
        let inline_distance = to_wall_origin.dot(wall.inline);

        if distance_to_wall >= 0.0 && distance_to_wall_previous < 0.0 && inline_distance.abs() < wall.length / 2.0 {
            return true;
        }
        return false;
    }

    pub fn has_hit_any(&self, walls: Vec<Wall>, act_on_both_sides: bool) -> bool {
        for wall in walls {
            if self.has_hit(wall, act_on_both_sides) {
                return true;
            }
        }
        return false;
    }

    pub fn has_hit_any_reverse(&self, walls: Vec<Wall>) -> bool {
        for wall in walls {
            if self.has_hit_reverse(wall) {
                return true;
            }
        }
        return false;
    }

    // pub fn new(init_pos: Vec2, radius: f64) -> Self {
    //     Ball {
    //         previous_position: vec2::ZERO,
    //         position: init_pos,
    //         radius
    //     }
    // }
}